### DISCUSSION TITLE
ELKS 0.8.0-pre2 on ESP32 FabGL/PCEmulator

### DISCUSSION THREAD

**Discussion by tyama501** [jbruchon/elks#1939] [Show and tell]
The latest ELKS 0.8.0-pre2 (commit 511e594 fd1440-fat.img for IBM) is booted on the ESP32 FabGL/PCEmulator.

![IMG_20240720_165228](https://github.com/user-attachments/assets/e7dfa632-7725-4921-8bc5-8b777540d4fd)

Many thanks to the developer of the FabGL for developing the great libraries and the emulator.
https://github.com/fdivitto/FabGL

The cpu core and the bios are based on the 8086tiny.
These are also great.

**Comment by ghaerr** [comment]
Wow @tyama501, that's pretty amazing seeing ELKS run on an ESP32 board. How is disk implemented, as USB flash? Is USB implemented in an embedded OS on the ESP32 board?

I see FabGL is primarily a graphics library and that VGA is emulated. Have you tried running nano-X programs on FabGL? I am wondering whether the VGA device is emulated in software or hardware.

> The cpu core and the bios are based on the 8086tiny.

The 8086tiny bios is pretty cool in that it has callouts implemented for host routines (like console I/O), and is well done. I remember that portions of the 8086tiny CPU emulator (e.g. instruction decoding) are also handled in an unusual fashion through BIOS code outside the C 8086 emulator. Is FabGL using the same mechanism? It would be interesting to see how well a recent version of ELKS runs on 8086tiny, that was the first emulator I used when I joined ELKS years ago.

**Reply by tyama501** [reply]
It is the same system in our readme.
(I just bought a board from the other developer and tested the latest elks)
https://github.com/ghaerr/elks?tab=readme-ov-file#watch-elks-in-action

I haven't tried the nano-X yet.
I will try it later.
(EDIT: Maybe it supports VGA)

I tested the mouse on the FreeDOS and it worked well but it support only PS/2 mouse.
We don't have PS/2 driver yet, right?

Disk images are on the micro SD card and selectable when booting.
WIFI is also available for getting image and setting the time for the RTC from the network but I haven't set it yet.

Yes, it seems that the cpu core and the bios codes are modified from the originality tiny, but the bios is still separated and compiled with NASM.

**Reply by tyama501** [reply]
The keyboard is connected with USB plug but it is also PS/2.
The keyboard works well.

**Reply by tyama501** [reply]
The comments says the decode table is moved from BIOS to C code.
https://github.com/fdivitto/FabGL/blob/master/src/emudevs/i8086.cpp

**Reply by ghaerr** [reply]
> We don't have PS/2 driver yet, right?

Nano-X has the PS/2 mouse data stream decoder but currently only reads from a serial port. There are no kernel mouse drivers in ELKS, but a PS/2 user-mode driver should be pretty easy using the PS/2 mouse interrupt in a modified nano-X mouse driver.

> (EDIT: Maybe it supports VGA)

You're right, Nano-X is currently compiled with only the CGA driver. A VGA driver is available but will need compilation into Nano-X and testing.

> The comments says the decode table is moved from BIOS to C code.

Interesting, thanks for the link.

**Reply by tyama501** [reply]
Hi @ghaerr ,

It seems that the current nx application (nxworld, nxtetris) does not work with black screen and cursor blinking.
It can go back to console with CTRL-C so the system is not crashed.

BTW,
I tried SL.
I think the speed is similar to the real 8086.

https://github.com/user-attachments/assets/8ae3899f-6a8c-4a4c-80bf-01d99f9225e8

**Reply by tyama501** [reply]
I don't think this is related to the emulator, but is TZ variable in etc/profile working?

TZ=JTC-9 works if I modify it in the bootopts,
but I couldn't set it in the etc/profile.
(I usually don't use this since PC-98 currently set TZ when building elks)

**Reply by ghaerr** [reply]
> TZ=JTC-9 works if I modify it in the bootopts,
> but I couldn't set it in the etc/profile.

Are you using `export TZ=JTC-9`? Without export it won't work.

**Reply by tyama501** [reply]
Oh, I see! thanks.

**Reply by tyama501** [reply]
Yes, the export worked. Thank you.

**Reply by tyama501** [reply]
It seems that the graphic library supports VGA but the pc emulator only supports cga video modes.
That is why the current EGA/VGA nano-X driver doesn't work.

int10
https://github.com/fdivitto/FabGL/blob/master/examples/VGA/PCEmulator/bios.asm

**Reply by tyama501** [reply]
Hello @ghaerr 

Currently there is no cga driver for elks nano-X right?
I may try to make one for
int10_set_vm_11:
  ; 640x480, 2 colour

**Reply by ghaerr** [reply]
No, but that would be great for ELKS if you wrote a CGA driver. It should be quite a bit simpler than the EGA driver.

> 640x480, 2 colour

Is that specific for PC-98? I thought two-color mode was 640x200. It might be nice to support 320x200 4-color also.

**Reply by tyama501** [reply]
https://github.com/fdivitto/FabGL/blob/master/examples/VGA/PCEmulator/machine.cpp

Oops
it seems that machine.c
only supports up to 640x200 vm6.

I am not sure it can scale to vga so that my LCD monitor can display it but maybe I will try it.

**Comment by tyama501** [comment]
fdisk, mkfat, and mounting 250MByte hard disk image.

![fdisk](https://github.com/user-attachments/assets/1baa6ffc-4299-4849-9cd7-85a2cbac89c0)

![mkfat](https://github.com/user-attachments/assets/9f03afef-9ab4-499b-9361-0d680fb6aafd)

**Comment by tyama501** [comment]
ttyclock
https://github.com/user-attachments/assets/50a68c7e-e64e-4c07-9f27-e993fcbd3051

It seems that redrawing once in a several seconds.

**Reply by ghaerr** [reply]
Hmm, I don't know whether that's because of `ttyclock` itself or just the total emulation of both 8086 kernel and app and lots of terminal output. Perhaps some other timing can be done that is not using extensive console output. I didn't think ttyclock is that slow on real 8086 though.

**Reply by tyama501** [reply]
It seems that ttyclock on i486 25MHz pc-98 emulator sometimes skips a couple of seconds,
so the several seconds in 8086 is not so bad.

Thank you.

**Comment by tyama501** [comment]
Hello @ghaerr ,

I made monochrome CGA 640x200 nano-X driver.
scr_cga.c and vgplane4_cga.c
https://github.com/tyama501/elks/commit/0c92b22d7639b83ea704e9c5666db8ae34e68a5f

I need to use ramfont instead of romfont.
DRIVERS += drivers/scr_cga.o drivers/vgaplan4_cga.o drivers/mempl4.o
DRIVERS += drivers/ramfont.o drivers/X6x13.o drivers/elksutilasm.o

I also made scr_cga.c to pretend as colored display so that current application can draw for now.
	//psd->bpp = 1;
	//psd->ncolors = 2;
    // To run colored applications
	psd->bpp = 4;
	psd->ncolors = 16;

FabGL/PCEmulator could draw it!
(Although the size of nxtetris is exceeding CGA so the bottom cannot be seen.)

nxtetris
![nxtetris_cga](https://github.com/user-attachments/assets/11a825c7-49aa-4763-bf2d-8a73f888664b)

nxworld
![nxworld_cga](https://github.com/user-attachments/assets/96885a39-27e8-4e18-a104-206cec9248ef)

nxterm
![nxterm_cga](https://github.com/user-attachments/assets/7fe28bfe-387b-4da6-b183-5165a77b3234)

nxlandmine
![nxlandmine_cga](https://github.com/user-attachments/assets/887c42fa-c679-40b5-8fa7-8dd8bbbb3b18)

For nxlandmine,
it seems that psd->ncolors = 2 has better result according to DOSBox-X,
but I don't care it since I don't have a serial mouse for the device.

<img width="470" alt="nxlandmine_2color" src="https://github.com/user-attachments/assets/b9d17ca3-9456-48a5-8d05-ed2e8fc02992">

Thank you!

**Reply by ghaerr** [reply]
Wow @tyama501, you sure wrote that screen driver quickly, and very impressive results!!!

Thanks for the screen shots. Sounds like we may need to think a bit about \#defines for game height/width sizes when compiled for different screen resolutions.

The nano-X compiles themselves could use a new variable setting the desired screen driver/resolution etc combination. This is what later versions of Nano-X do, but this doesn't need to be nearly as complicated. An ELKS .config file option could also be used, I suppose.

> (Although the size of nxtetris is exceeding CGA so the bottom cannot be seen.)

The games could have SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_NCOLORS etc from a Nano-X driver .h file included in nano-X.h that override their default setting according to the compiled-in screen driver. You're welcome to go ahead and make changes in that fashion.

Thank you!

**Reply by toncho11** [reply]
So @tyama501  this new cga driver will allow vga output to kind of work in ibm PCs with cga monitors?

**Reply by tyama501** [reply]
Hi @toncho11 

I think the FabGL is scan converting from the CGA to VGA since my LCD monitor does not allow 640x200 interlace input.
So if you can test it with the real CGA monitor, it would be very nice.

**Reply by tyama501** [reply]
I have added getenv "MONOMODE" to the driver.
https://github.com/tyama501/elks/commit/dfb8de67fb6c633f27de67528d2cd8f29960f698

When export MONOMODE, it quit pretending 16 colors and psd->ncolors get 2, so the nxlandmine can be drawn.
	if (MONOMODE) {
		psd->bpp = 1;
		psd->ncolors = 2;
	} else {
		// To run colored applications
		psd->bpp = 4;
		psd->ncolors = 16;
	}

For nxtetris,
it seems that if we modify WELL_HEIGHT to 18, it is playable.
(althougth key instruction is out of screen)

ntetris.h
//#define WELL_HEIGHT 28
//#define WELL_VISIBLE_HEIGHT 24
#define WELL_HEIGHT 18
#define WELL_VISIBLE_HEIGHT 16

<img width="362" alt="nxtetris_cga" src="https://github.com/user-attachments/assets/2a280a20-257a-4e91-af9d-b8417e03d535">

Can we have some ifdef CGA definition from configuration?
If I can get one I will also add it to the make file.

Thank you.

**Reply by tyama501** [reply]
Hello @ghaerr 

BTW
I have been looking the nxtetris source code and "d" for right rotate is not correct.
I think the right rotate is key "f".
<img width="371" alt="image" src="https://github.com/user-attachments/assets/4c0c1ff3-e933-4923-945a-efce48c563f6">

                case 'd':
                case 's':
                case 'D':
        /*case MWKEY_UP:*/
                        rotate_block(state, 0);
                        break;
                case 'f':
                case 'g':
                case 'F':
        /*case MWKEY_DOWN:*/
                        rotate_block(state, 1);

**Reply by ghaerr** [reply]
Yes - it appears the help display is incorrect. I suppose the display should show s and f, or perhaps d for left and f for right, if it is wanted to keep the keys next to each other like j and k are....

**Reply by tyama501** [reply]
Hello @ghaerr ,

I see there is CONFIG_HW_VGA in config.
elks/arch/i86/boot/setup.S:#ifdef CONFIG_HW_VGA
elks/arch/i86/boot/setup.S:#endif /* CONFIG_HW_VGA*/
elks/arch/i86/defconfig:CONFIG_HW_VGA=y
elks/config.in:         bool 'VGA adapter'      CONFIG_HW_VGA       y

Can we use ifndef CONFIG_HW_VGA as CGA
for the makefile and the application?

**Reply by tyama501** [reply]
I mean not CONFIG_ARCH_PC98 case.

**Reply by ghaerr** [reply]
> Can we use ifndef CONFIG_HW_VGA as CGA
for the makefile and the application?

Yes, that sounds like a good idea!

**Reply by tyama501** [reply]
OK, I will try and make PR if it works.
Thanks.

**Reply by tyama501** [reply]
Hello @ghaerr 

I forgot about EGA.

For EGA PC
VGA driver need to be included without the CONFIG_HW_VGA right?
Then export=EGAMODE set the VGAMODE as FALSE.

https://github.com/ghaerr/elks/blob/master/elkscmd/nano-X/drivers/scr_bios.c
	if(getenv("EGAMODE"))
		VGAMODE = FALSE;
	else VGAMODE = TRUE;

We may need new definition CONFIG_HW_CGA (or CONFIG_HW_EGA) to distinguish these.

**Reply by ghaerr** [reply]
> We may need new definition CONFIG_HW_CGA (or CONFIG_HW_EGA) to distinguish these.

I would rather not go that route since the meaning of CONFIG_HW_VGA is actually about whether the BIOS supports a call to return the screen resolution, and the information isn't used by the kernel in the original way anymore. Ultimately that config variable should be going away.

> VGA driver need to be included without the CONFIG_HW_VGA right?

Yes, for both VGA and EGA (=not CGA) , and it seems that setting EGAMODE in that case allows EGA resolution for compiled nano-X applications.

If CONFIG_HW_VGA is not set, the CGA driver will be included. This would mean that neither VGA or EGA is supported. Will that work, or is there another problem I am not understanding?

**Reply by tyama501** [reply]
Oh, OK. So, it means currently CONFIG_HW_VGA can be set for EGA PC too. (since it isn't used in the original way)

**Comment by lmpp1972** [comment]
Hi @tyama501
Thank you for pointing me to your work, i have download anyway and tried to run, but  my IBM original Hercules card show no color or graphics only a blinking "_" , only thing i can run is matrix!!

**Reply by tyama501** [reply]
Hi @lmpp1972 

Thank you for testing it!
Oh, I see.

Hi @ghaerr and @toncho11 
Do you know something more to work it with the real hardware?
Or do you think the card just don't support it?

Thank you.

**Reply by ghaerr** [reply]
I think the problem @lmpp1972 is having here is that the Hercules card might not be CGA compatible, and possibly uses a different address for video memory, so nothing is displayed. Thus @tyama501's new CGA driver won't work for it. IIRC Hercules was a mono-only card that supported some better text resolutions only?

**Reply by tyama501** [reply]
Thank you @ghaerr 

It seems MDA with bitmap graphic, but different from CGA.
https://en.wikipedia.org/wiki/Hercules_Graphics_Card

**Comment by tyama501** [comment]
Hello @gharr

I have a question about BIOS call of IBM PCs.
I haven't found much BIOS calls that can be used to implement graphic fuctions for ELKS basic but does IBM PC have BIOS calls to draw images? Or do we need to draw directly to vram?
Well actually there are some in 
int10 but not so many as PC-98 has.

**Reply by ghaerr** [reply]
Hi @tyama501,

Yes there are some BIOS calls to display graphics but in general they are very slow and not recommended for use in games. Unfortunately for the computer vintages we're concerned with here, that means different drivers for various cards, the popular being VGA, EGA, CGA, etc and possibly Hercules or other very old cards. These older cards also had the 20 bit memory addressing limitations of the real mode CPU, and thus had extra registers to select memory planes for reading/writing.

On newer systems that supported 32-bit (protected) mode CPUs, the video cards could use linear "framebuffer" memory for much speedier access somewhere usually high in the physical address space, but virtually none of these cards are usable on ELKS, which is why Nano-X has multiple graphics drivers.

Early PC BIOSes implemented some very basic read/write pixel calls in order to provide some mechanism that would allow for a graphics display to be presented without the programmer having to worry about which card (and video mode) might be present and in use. As such, most BIOS implementations for graphics requests are nowhere near fast enough for animation, but were sometimes used in non-animated graphics display. Of course, all this can be tested and played with for speed, but look at the Nano-X EGA driver: it already takes quite a bit of code just to read/write a byte of display memory, so adding interrupt processing overhead and different mode handling is not desirable.

**Reply by tyama501** [reply]
Starting to make host-cga.c and add host_plot and host_draw for graphic mode 05, 320x200 with int10.

host-cga.c
https://github.com/tyama501/elks/commit/292165b58e1a38ff06fbe40034071978e1f7e86f

It seems that the host_draw still needs debugging a lot.
![image](https://github.com/user-attachments/assets/bc604a47-7817-4f0c-9079-21777456a5ef)

**Reply by ghaerr** [reply]
Hello @tyama501,

That's pretty cool using an INT 10h BIOS call for CGA graphics in basic!

The inner loops of your host_draw routines seem overly complicated, which will slow them down. Also, problem may be related to not keeping an bresenham "error" coefficient that needs adding back in. See nano-X/engine/devdraw.c::GdLine() for how Nano-X does it (it's code is a bit overly complicated due to the engine having to worry about whether the last point is drawn or not, for incompatibility between Win32 and X11 drawing).

I'm not sure what the other background issue(s) are, but from looking at the "ELKS" characters, it seems the x/y coordinates may be getting doubled somewhere, or is that required for CGA coordinates?

Thank you!

**Reply by tyama501** [reply]
Hello @ghaerr ,

Thank you!
I will look it and debug it when I have more time.

>it seems the x/y coordinates may be getting doubled somewhere

This is because my original test98.bas is doubling it and it is intended :)

**Reply by tyama501** [reply]
Hi @ghaerr ,

I haven't added bresenham "error" coefficient, but it got better.
https://github.com/tyama501/elks/commit/d350f0aa356ee7e077f5a39bcefee88a4ecbcbc4

![basic_test1](https://github.com/user-attachments/assets/0a5f3ad1-a4f3-43d1-9c30-daaa7a517316)


![basic_test1_run](https://github.com/user-attachments/assets/4d101461-8c03-40f5-abe0-d6669398a734)


![basic_testcga](https://github.com/user-attachments/assets/864c5381-40ee-4b08-b2d6-aa80d5d0aded)


![basic_testcga_run](https://github.com/user-attachments/assets/e9ac8ce9-b18c-4092-8455-90890358fc83)

CLS won't work yet.

Also, unlike PC-98
we cannot print text while graphic mode.

So, if we forget to add MODE 0 before end or stop program with CTRL-C,
then the text will never be shown.
How this should be?

Thank you.

**Reply by ghaerr** [reply]
> I haven't added bresenham "error" coefficient, but it got better.

Looking very nice! Test line draws at other than 45 degrees to see how the drawing varies on the X vs Y axis to test your algorithm.

> if we forget to add MODE 0 before end or stop program with CTRL-C,
then the text will never be shown.

I think the simplest way would be to register a pre-exit handler routine with `atexit` in the host_mode routine. This would ensure that the special exit pre-handler would get called only if a mode command was previously executed. The exit handler could then check the saved last mode value and reset to text mode if required. This approach keeps everything only within the host*.c files, rather than in the basic interpreter, and automatically resets to text mode only when needed. (Note that atexit should only be called once to register a single handler, not multiple times, just in case the mode command is executed many times).

**Reply by tyama501** [reply]
Hello @ghaerr 

By introducing 7bits fixed binary fraction,
I could draw line other than 45 degrees.

https://github.com/tyama501/elks/blob/basic_cga/elkscmd/basic/host-cga.c

![basic_test2](https://github.com/user-attachments/assets/51c86c88-6df2-42af-ab78-581fe5e022fe)

![basic_test2_run](https://github.com/user-attachments/assets/6024590b-30e9-4395-b920-ca82bfcd1df3)

Here is the movie drawing on the FabGL/PCEmulator.
(It seems the default palette is different)

https://github.com/user-attachments/assets/2416cfa4-1401-412c-b399-1784fc07f35e

I think the drawing speed is not bad for now.
I will think about moving on to the bresenham algorithm if there are strong request for speed up.

I haven't implemented the atexit and CLS yet.

Thank you!

**Reply by ghaerr** [reply]
That looks great, thanks for the images!

With regards to speed, it's usually the area-fill that gets noticed when operating too slowly, compared to the line draws which are relatively quite fast.

**Reply by tyama501** [reply]
Hi @ghaerr ,

I have a couple of questions of the atexit for the basic.

Where should I put the atexit? 
(Sorry I couldn' t fully understand https://github.com/ghaerr/elks/discussions/1939#discussioncomment-10440405)
Currently host_mode is setting mode.
https://github.com/tyama501/elks/blob/basic_cga/elkscmd/basic/host-cga.c#L74

Do I need to add some kind of handler like host_exit that calls host_mode
and assign the host_exit with atexit(host_exit)?

Or host_exit does not call host_mode but it calls int10 directly.
Then atexit(host_exit) can be in the host_mode.
Is this your idea?

One more question,
does the exit work
not only for CTRL-D (exit basic),
but also, for CTRL-C (quit program)?

For CLS command
I am thinking to clear all CGA VRAM with host_pokeb.

Thank you.

**Reply by ghaerr** [reply]
Hello @tyama501,

Sorry, I have been in a deep dive trying to understand a major bug in QEMU running large ELKS or TLVC kernels. Looks like we're getting somewhere and then I will have more time :)

> Where should I put the atexit?

The atexit should go in the host_mode routine, and be called just once, the first time a graphics mode is requested.

The way atexit works, is you pass it a routine that the C library ensures will be called whenever exit() (but not _exit()) is called for any reason.

> does the exit work
not only for CTRL-D (exit basic),
but also, for CTRL-C (quit program)?

Yes, when basic exits with ^D, it calls exit(), so the atexit handler will run.
If basic exits on a ^C through an interrupt handler, then if that interrupt handler calls exit, the atexit handler will run the registered routines. If _exit() is called instead, they won't. Basic should only be calling exit(), as _exit is for special purposes.

It would be possible to also have the ^C SIGINT handler call host_mode() to set back text mode. This would then make ^C exit out of graphics mode, which might (or might not) be good for users.

> Or host_exit does not call host_mode but it calls int10 directly.
Then atexit(host_exit) can be in the host_mode.
Is this your idea?


Yes - host_exit will be called by atexit because host_mode called atexit(host_exit) (only if graphics mode is turned on for the first time though). Then, when host_exit is called as a result of graphics mode having been on once, it checks to see whether graphics mode is still on, and makes the INT 10 call to go back to text mode.

I hope that clears things up.

> For CLS command
I am thinking to clear all CGA VRAM with host_pokeb.

That would work, but a fast method would be to use an `fmemset` which takes a far pointer destination and clears it to zeros. We have fmemset in the kernel under elks/arch/i86/lib/fmemory.S if you want to take a look at how that one works.

Thank you!

**Reply by tyama501** [reply]
Hello @ghaerr 

I have referred to the fmemset and wrote clearing codes of vram by in-line assembler.
CLS worked well.
https://github.com/tyama501/elks/commit/c97b34955a2a5d296bc6911b268a2a024d90b378

Also, I have added the atexit.
CTRL-D exit worked well. CTRL-C does not work, but that can be next step.
https://github.com/tyama501/elks/commit/da12e765470e5038835e63cad351c015d0a3885b

I modifed snake.bas a little to snakcga.bas to eliminate CIRCLE command and switching text mode and video mode.
(This still need debug)
https://github.com/tyama501/elks/commit/0c1001e93bd1b6b76f86d52da0967e61fec04afe

https://github.com/user-attachments/assets/c4cdbadb-b048-4cbe-a381-f4f15cb47f97

Well, I think I would like to make first PR for host-cga.c.
How should I do for the Makefile?
Do I need to add some ifdef?
or is it ok to replace host-stubs.c.

Thank you!

**Reply by ghaerr** [reply]
Hello @tyama501,

Looks nice!

> hink I would like to make first PR for host-cga.c.
How should I do for the Makefile?
Do I need to add some ifdef?
or is it ok to replace host-stubs.c.

The `basic` executable is now built for IBM PC, PC-98, and also for testing for the Watcom library build as well as for the host using `make hostbasic`. Now that you have built support for CGA on IBM PC, we need another host*.c file, since the others are already being used separately and host-stubs is important to keep for when having no special machine features.

I would suggest that you rename host-cga.c to host-ibmpc.c since it is for the IBM PC platform, even though only CGA support for now. All IBM PC-specific code can go in there. Then change the Makefile so that the ELKS IBM PC build uses host-ibmpc.o instead of host-stubs.o. There are still issues as to which compiler is being used for IBM PC, see below.

> Also, I have added the atexit.

Looks good, exactly what I was thinking. ^C handling should be pretty straightforward, calling the mode reset in the signal handler if needed. Since `catchint` is in host.c, you may need to either rename that and call it host_catchint, or leave it the same and perform a host_catchint() callout from the catchint routine. This would involved adding host_catchint to all host*.c files to keep the build running.

> I have referred to the fmemset and wrote clearing codes of vram by in-line assembler.

It looks like the automatic OpenWatcom CI routine in .github/worflows/ow-libc.yml uses host-stubs.c for building, so we don't need to worry yet about GCC-specific coding, but you should plan on any non portable C to use either `#if __ia16__` or `#ifndef __WATCOMC__` so that both compile and work. It is too bad that there is not a common C ASM or ASM file compatibility between the two compilers. So for now we have to use ifdef. IMO it is sometimes it is better to the `#ifndef __WATCOMC__` since it shows that Watcom C is not implemented, rather than the reverse for ia16, where we always assume that we have ia16-elf-gcc to work with.

Thank you!
